src/k_means_cluster.js

```diff
@@ -0,0 +1,123 @@
+ */
+const kMeansCluster = (points, numCluster, randomSource=Math.random) => {
+    let oldCentroids = null,
+        newCentroids = generateInitialCentroids(points, numCluster, randomSource),
+        labels = null,
+        change = Number.MAX_VALUE;
+    while (change !== 0) {
+        labels = labelPoints(points, newCentroids);
+        oldCentroids = newCentroids;
+        newCentroids = calculateCentroids(points, labels, numCluster);
+        change = calculateChange(newCentroids, oldCentroids);
+    }
+    return {labels: labels, centroids: newCentroids};
+}
+
+/**
+ * Generate starting points for clusters by randomly selecting points.
+ * @param {Array<Array<number>>} points Array of XY coordinates.
+ * @param {number} num How many points to select.
+ * @param {function} randomSource Generate uniform random values in [0, 1).
+ * @returns {Array<Array<number>>} XY coordinates of centroids.
+ */
+const generateInitialCentroids = (points, num, randomSource) => {
+    return sample(points, num, randomSource);
+}
+
+/**
+ * Label each point according to which centroid it is closest to.
+ * @param {Array<Array<number>>} points Array of XY coordinates.
+ * @param {Array<Array<number>>} centroids Current centroids.
+ * @returns {Array<number>} Group labels.
+ */
+const labelPoints = (points, centroids) => {
+    return points.map(p => {
+        let minDist = Number.MAX_VALUE,
+            label = -1;
+        centroids.forEach((c, i) => {
+            const dist = distance(p, c);
+            if (dist < minDist) {
+                minDist = dist;
+                label = i;
+            }
+        })
+        return label;
+    })
+}
+
+/**
+ * Calculate centroids for points given labels.
+ * @param {Array<Array<number>>} points Array of XY coordinates.
+ * @param {Array<number>} labels Which groups points belong to.
+ * @param {number} num Number of clusters being created.
+ * @returns {Array<Array<number>>} Centroid for each group.
+ * @throws {Error} If any centroids wind up friendless (i.e., without associated points).
+ */
+const calculateCentroids = (points, labels, num) => {
+    // Initialize accumulators.
+    const width = points[0].length;
+    const centroids = [];
+    for (let i=0; i<num; i++) {
+        centroids.push(Array(width).fill(0));
+    }
+    const counts = Array(num).fill(0);
+
+    // Add points to centroids' accumulators and count points per centroid.
+    points.forEach((p, i) => {
+        const lbl = labels[i];
+        const current = centroids[lbl];
+        current.forEach((soFar, j) => {
+            current[j] = soFar + p[j];
+        })
+        counts[lbl] += 1;
+    })
+
+    // Rescale centroids, checking for any that have no points.
+    centroids.forEach((c, i) => {
+        if (counts[i] === 0) {
+            throw new Change(`Centroid ${i} has no friends`);
+        }
+        centroids[i] = centroids[i].map(total => total / counts[i]);
```

Comment:
Let's do a traditional loop and overwrite the values here - the `.map` is going to allocate a new array per iteration.

Reply:
:+1: